core/src/main/java/org/apache/calcite/rel/metadata/RelMdColumnUniqueness.java

```diff
@@ -285,21 +287,86 @@ public Boolean areColumnsUnique(Intersect rel, RelMetadataQuery mq,
 
     // for FULL OUTER JOIN if columns contain column from both inputs it is not
     // guaranteed that the result will be unique
     if (!ignoreNulls && rel.getJoinType() == JoinRelType.FULL
-        && leftColumns.cardinality() > 0 && rightColumns.cardinality() > 0) {
+        && initLeftColumns.cardinality() > 0 && initRightColumns.cardinality() > 0) {
       return false;
     }
 
+    // If we have an inner join on either side, we may be able to determine
+    // uniqueness from the condition.
+    HashSet<RexCall> equalities = findCommonEqualities(rel.getCondition());
+    // Check each equality. If we are only checking uniqueness with a single
+    // side we can include the other side.
+    BitSet addedLeftCols = new BitSet();
+    BitSet addedRightCols = new BitSet();
+    boolean isOuterLeft = rel.getJoinType().generatesNullsOnRight();
+    boolean isOuterRight = rel.getJoinType().generatesNullsOnLeft();
+    for (RexCall equals: equalities) {
+      Integer arg0 = ((RexInputRef) equals.getOperands().get(0)).getIndex();
+      Integer arg1 = ((RexInputRef) equals.getOperands().get(1)).getIndex();
+      if (arg0 < numLeftCols && arg1 < numLeftCols) {
+        // Both columns are in the left table.
+        if (!isOuterLeft) {
+          // An outer join may not have the columns equal
+          boolean foundArg0 = initLeftColumns.get(arg0);
+          boolean foundArg1 = initLeftColumns.get(arg1);
+          if (foundArg0 && !foundArg1) {
+            addedLeftCols.set(arg1);
+          } else if (!foundArg0 && foundArg1) {
+            addedLeftCols.set(arg0);
+          }
+        }
+      } else if (arg0 >= numLeftCols && arg1 >= numLeftCols) {
+        // Both columns are in the right table.
+        if (!isOuterRight) {
+          // An outer join may not have the columns equal
+          arg0 = arg0 - numLeftCols;
+          arg1 = arg1 - numLeftCols;
+          boolean foundArg0 = initLeftColumns.get(arg0);
+          boolean foundArg1 = initLeftColumns.get(arg1);
+          if (foundArg0 && !foundArg1) {
+            addedRightCols.set(arg1);
+          } else if (!foundArg0 && foundArg1) {
+            addedRightCols.set(arg0);
+          }
+        }
+      } else {
+        // Left and right are in different tables
+        Integer leftArg;
+        Integer rightArg;
+        if (arg0 < numLeftCols) {
+          // arg0 -> Left, arg1 -> Right
+          leftArg = arg0;
+          rightArg = arg1 - numLeftCols;
+        } else {
+          // arg0 -> Right, arg1 -> Left
+          leftArg = arg1;
+          rightArg = arg0 - numLeftCols;
+        }
+        boolean foundLeft = initLeftColumns.get(leftArg);
+        boolean foundRight = initRightColumns.get(rightArg);
+        if (foundLeft && !foundRight && !isOuterLeft) {
+          addedRightCols.set(rightArg);
+        } else if (!foundLeft && foundRight && !isOuterRight) {
+          addedLeftCols.set(leftArg);
+        }
```

Comment:
ditto

Reply:
I restructured this code to get outer, so I'll keep this the same.